x-pack/legacy/plugins/code/server/distributed/cluster/node_repositories_service.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import util from 'util';
+import { ClusterService, ClusterStateListener } from './cluster_service';
+import { ClusterStateEvent } from './cluster_state_event';
+import { ClusterMembershipService } from './cluster_membership_service';
+import { CloneWorker } from '../../queue';
+import { Repository, RepositoryUri } from '../../../model';
+import { Logger } from '../../log';
+
+export class NodeRepositoriesService implements ClusterStateListener {
+  // visible for test
+  readonly localRepos = new Map<RepositoryUri, LocalRepository>();
+  private readonly localNodeId = this.clusterMembershipService.localNode.id;
+
+  constructor(
+    private readonly log: Logger,
+    private readonly clusterService: ClusterService,
+    private readonly clusterMembershipService: ClusterMembershipService,
+    private readonly cloneWorker: CloneWorker
+  ) {}
+
+  public async start() {
+    /**
+     * we can add locally exists repositories to localRepos when the service is started to avoid unnecessarily add clone
+     * tasks for them, but for now it's OK because clone job is idempotent.
+     */
+    this.clusterService.addClusterStateListener(this);
+  }
+
+  public async stop() {}
+
+  async onClusterStateChanged(event: ClusterStateEvent): Promise<void> {
+    // compare repositories in the cluster state with repositories in the local node, and remove
+    const repos = event.current.getNodeRepositories(this.clusterMembershipService.localNode.id);
+    const localNewRepos = repos.filter(repo => !this.localRepos.has(repo.uri));
+    const localRemovedRepos = Array.from(this.localRepos.values()).filter(
+      repo =>
+        event.current.routingTable.getNodeIdByRepositoryURI(repo.metadata.uri) !== this.localNodeId
+    );
+    for (const localNewRepo of localNewRepos) {
+      this.log.info(
+        `Repository added to node [${this.localNodeId}]: ${util.inspect(localNewRepo)}`
+      );
+      await this.cloneWorker.enqueueJob({ url: localNewRepo.url }, {});
+      this.localRepos.set(localNewRepo.uri, {
+        metadata: localNewRepo,
+        currentState: RepositoryState.CLONING,
+      });
+    }
+    for (const localRemovedRepo of localRemovedRepos) {
+      this.log.info(
+        `Repository removed from node [${this.localNodeId}]: ${util.inspect(
+          localRemovedRepo.metadata
+        )}`
+      );
+      this.localRepos.delete(localRemovedRepo.metadata.uri);
```

Comment:
do we need to some additional steps to clean up the repository in the filesystem? 

Also, since we won't have the repo size load auto-balancing in the first version or in the near future, I am just wondering what would be the case when a repo can be removed from a node because of cluster state change? A node failes and then comes back to live again?

Reply:
Sure, this is a problem, and I would like to leave it to future PRs, because that involves gathering and maintaining the states of local repositories and need more work and discussion.

By the way, once we have the implementation to clean up the repository based on cluster state change, it could also be used to delete the repository from the cluster: we can just mark the repository as 'deleting' in ES, the node where the repository resides will see the change and found the repository doesn't belongs to it any more and delete the repo from it. And if the clean up is triggered by the action of 'deleting', then the node can delete related indices from ES(or we can do this at the very beginning?) and mark the repository as 'deleted'(or whatever we do now).

If we do it like this, I think the cluster mode is evolving towards a cluster state driven implementation.

Reply:
maybe open an issue tracking this and implement later when  kibana HA is ready

Reply:
@fantapsody it may make sense to add a TODO comment here.